serverless/aws/protomaps-template.yaml

```diff
@@ -43,94 +46,87 @@ Resources:
               Service: lambda.amazonaws.com
             Action: sts:AssumeRole
       Policies:
-        - PolicyName: LambdaBasicExecution
+        - PolicyName: !Sub "${AWS::StackName}-LambdaLoggingPolicy"
           PolicyDocument:
             Version: '2012-10-17'
             Statement:
               - Effect: Allow
                 Action:
-                  - logs:CreateLogGroup
                   - logs:CreateLogStream
                   - logs:PutLogEvents
-                Resource: '*'
-        - PolicyName: S3AccessPolicy
+                Resource: !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/${AWS::StackName}:log-stream:*"
+        - PolicyName: !Sub "${AWS::StackName}-LambdaS3AccessPolicy"
           PolicyDocument:
             Version: '2012-10-17'
             Statement:
               - Effect: Allow
                 Action: s3:GetObject
                 Resource: !Sub arn:aws:s3:::${BucketName}/*
 
-##########################################################################
-#   Lambda Function                                                      #
-##########################################################################
-
-  ProtomapsLambdaFunction:
-    Type: 'AWS::Lambda::Function'
-    Properties:
-      FunctionName: protomaps
-      Runtime: nodejs18.x
-      Architectures: [arm64]
-      Role: !GetAtt LambdaExecutionRole.Arn
-      Handler: index.handler
-      MemorySize: 512
-      Environment:
-        Variables:
-          BUCKET: !Ref BucketName
-          PUBLIC_HOSTNAME: !Ref PublicHostname
-      Code:
-        S3Bucket: !Ref CodeBucketName
-        S3Key: !Ref CodeKey
 
-  ProtomapsLambdaFunctionUrl:
+  LambdaFunctionUrl:
     Type: 'AWS::Lambda::Url'
     Properties:
       AuthType: NONE
-      TargetFunctionArn: !GetAtt ProtomapsLambdaFunction.Arn
-      Cors:
-        AllowOrigins: ["*"]
+      TargetFunctionArn: !GetAtt LambdaFunction.Arn
       InvokeMode: BUFFERED
 
-  ProtomapsLambdaFunctionUrlPermission:
+  LambdaFunctionUrlPermission:
     Type: 'AWS::Lambda::Permission'
     Properties:
       Action: lambda:InvokeFunctionUrl
-      FunctionName: !Ref ProtomapsLambdaFunction
+      FunctionName: !Ref LambdaFunction
       Principal: '*'
       FunctionUrlAuthType: NONE
 
-# ##########################################################################
-# #  CloudFront::Distribution                                              #
-# ##########################################################################
+  ViewerRequestCloudFrontFunction:
```

Comment:
This is necessary in the case where `PublicHostname` parameter is left blank.

By default, the Lambda has no way to know what the CloudFront distribution's domain name is.

One way to do this is to set `PUBLIC_HOSTNAME` to reference the CloudFront resource in the CF template, but this introduces a circular dependency between the Lambda and the Distribution.

Some hacks to remove circular dependencies like listed in https://theburningmonk.com/2022/05/how-to-work-around-cloudformation-circular-dependencies/ such as creating the Lambda function ARN by hand don't work; the Lambda Function URL checks on provisioning the referenced function exists.

So instead, what we do is we check for the `PublicHostname` parameter, if blank we install a CloudFront Function that simply attaches the distribution domain name as the `x-distribution-domain-name` header that the Lambda can then read from.

Reply:
(The code for reading `x-distribution-domain-name` is not yet in this PR)

Reply:
The downside of adding CF Functions is that it costs $0.1 per million requests that go into CloudFront (very little)